src/d3a/models/strategy/external_strategies/pv.py

```diff
@@ -354,6 +364,71 @@ def _offer_aggregator(self, arguments):
 
 
+class PVForecastExternalStrategy(PVExternalMixin, PVPredefinedStrategy):
+    """
+        Strategy responsible for reading single production forecast data via hardware API
+    """
+    parameters = ('power_forecast_W', 'panel_count', 'initial_selling_rate', 'final_selling_rate',
+                  'fit_to_limit', 'update_interval', 'energy_rate_decrease_per_update',
+                  'use_market_maker_rate')
+
+    def __init__(
+            self, power_forecast_W: float = 0,
+            initial_selling_rate: float = ConstSettings.GeneralSettings.DEFAULT_MARKET_MAKER_RATE,
+            final_selling_rate: float = ConstSettings.PVSettings.SELLING_RATE_RANGE.final,
+            fit_to_limit: bool = True,
+            update_interval=duration(
+                minutes=ConstSettings.GeneralSettings.DEFAULT_UPDATE_INTERVAL),
+            energy_rate_decrease_per_update=None,
+            use_market_maker_rate: bool = False):
+        """
+        Constructor of PVForecastStrategy
+        :param power_forecast_W: forecast for the next market slot
+        """
+        super().__init__(panel_count=1,
+                         initial_selling_rate=initial_selling_rate,
+                         final_selling_rate=final_selling_rate,
+                         fit_to_limit=fit_to_limit,
+                         update_interval=update_interval,
+                         energy_rate_decrease_per_update=energy_rate_decrease_per_update,
+                         use_market_maker_rate=use_market_maker_rate)
+        self.power_forecast_buffer_W = power_forecast_W
+
+    @property
+    def channel_dict(self):
+        return {**super().channel_dict,
+                f'{self.channel_prefix}/set_power_forecast': self._set_power_forecast}
+
+    def _incoming_commands_callback_selection(self, req):
+        if req.request_type == "offer":
+            self._offer_impl(req.arguments, req.response_channel)
+        elif req.request_type == "delete_offer":
+            self._delete_offer_impl(req.arguments, req.response_channel)
+        elif req.request_type == "list_offers":
+            self._list_offers_impl(req.arguments, req.response_channel)
+        elif req.request_type == "device_info":
+            self._device_info_impl(req.arguments, req.response_channel)
+        elif req.request_type == "set_power_forecast":
+            self._set_power_forecast_impl(req.arguments, req.response_channel)
+        else:
+            assert False, f"Incorrect incoming request name: {req}"
+
+    def event_market_cycle(self):
+        self.produced_energy_forecast_kWh()
+        super().event_market_cycle()
+
+    def event_activate_energy(self):
+        self.produced_energy_forecast_kWh()
+
+    def produced_energy_forecast_kWh(self):
+        # sets energy forecast for next_market
+        energy_forecast_kWh = (self.area.config.slot_length / duration(hours=1)) / 1000 * \
+            self.power_forecast_buffer_W
```

Comment:
This calculation exists in many places now. How about refactoring to two utility functions called `convert_energy_to_power` and `convert_power_to_energy` respectively ?